File: evaluate/testuri.pl
use JSON;
use URI;

my %optionals = (
  hostname => host,
  username => user,
  password => password,
  port => _port
);

open TESTDATA, "urltestdata.json";
my $testdata = decode_json(join("", <TESTDATA>));
my @constructor_results=();

foreach (@$testdata) {
  my %test = %$_;
  my $uri = URI->new_abs($test{input}, $test{base})->canonical;

  my %result = (
    input => $test{input},
    base => $test{base},
    href => $uri->as_string,
    protocol => $uri->scheme . ":",
    pathname => $uri->path,
    search => $uri->can('query') && $uri->query ? '?'.$uri->query : '',
    hash => $uri->fragment ? '#'.$uri->fragment : ''
  );

  if ($uri->can('userinfo')) {
    my ($username, $password) = split /:/, $uri->userinfo, 2;
    $result{username} = $username;
    $result{password} = $password;
  }

  while (($var, $method) = each %optionals) {
    $result{$var} = $uri->$method if $uri->can($method);
  }

  push @constructor_results, \%result;
};

my %results = (
  useragent => 'Perl URI version ' . URI->VERSION,
  constructor => \@constructor_results
);

do {
  no warnings "utf8";
  binmode(STDOUT, ":utf8");
  print to_json(\%results, {pretty => 1});
}

File: evaluate/testuri.rb
require 'json'
require 'uri'

constructor_results = []

JSON.parse(File.read('urltestdata.json')).each do |test|
  result = {}

  begin
    uri = URI.parse test['input']
    uri = URI.join test['base'], uri
  rescue URI::InvalidURIError, URI::InvalidComponentError => e
    result[:exception] =  e.message
    uri = URI.parse('')
  end

  result.merge!({
    input: test['input'], 
    base: test['base'],
    href: uri.to_s,
    protocol: "#{uri.scheme}:",
    port: uri.port.to_s,
    hostname: uri.host,
    pathname: uri.path
  })

  result[:search] = "?#{uri.query}" unless uri.query.to_s.empty?
  result[:hash] = "##{uri.fragment}" unless uri.fragment.to_s.empty?

  constructor_results << result
end

useragent = `ruby -v`.chomp

puts JSON.pretty_generate(useragent: useragent, constructor: constructor_results)
